homeassistant/components/google_assistant/trait.py

```diff
@@ -1047,3 +1036,67 @@ def query_attributes(self):
                             ATTR_ENTITY_ID: self.state.entity_id,
                             media_player.ATTR_INPUT_SOURCE: source
                         }, blocking=True, context=data.context)
+
+
+@register_trait
+class OpenCloseTrait(_Trait):
+    """Trait to open and close a cover.
+    https://developers.google.com/actions/smarthome/traits/openclose
+    """
+
+    name = TRAIT_OPENCLOSE
+    commands = [
+        COMMAND_OPENCLOSE
+    ]
+
+    @staticmethod
+    def supported(domain, features):
+        """Test if state is supported."""
+        return domain == cover.DOMAIN
+
+    def sync_attributes(self):
+        """Return opening direction."""
+        return {
+            'openDirection': ['UP']
+        }
+
+    def query_attributes(self):
+        """Return state query attributes."""
+        domain = self.state.domain
+        response = {}
+
+        if domain == cover.DOMAIN:
+            position = self.state.attributes.get(cover.ATTR_CURRENT_POSITION)
+            if position is not None:
+                response['openPercent'] = position
+                response['on'] = self.state.state != cover.STATE_CLOSED
+                response['online'] = True
+            else:
+                response['on'] = self.state.state != cover.STATE_CLOSED
+                response['online'] = True
+                if self.state.state != cover.STATE_CLOSED:
+                    response['openPercent'] = 100
+                else:
+                    response['openPercent'] = 0
+
+        return response
+
+    async def execute(self, command, data, params):
+        """Execute an Open, close, Set position command."""
+        domain = self.state.domain
+
+        if domain == cover.DOMAIN:
+            if params['openPercent'] == 0:
```

Comment:
Also we probably don't need to route 'openPercent' 100% and 0% to open and close service calls respectively as they just forward to the set position service anyway.

Reply:
We need to route 0% and 100% to open and close service because google assistant always pass 'openPercent' even if no percentage is specified. It's a problem for covers that do not support set_cover_position service.

Reply:
Does the trait support blinds/garage doors that do not support setting a position?

Reply:
Yes it does support garage doors, that's the purpose of these conditions. Without routing 0% and 100% to close and open services the trait can't control garage doors. However if a percentage is specified in the voice command, it will call the set_cover_position service even if the cover doesn't support it and the door won't move.

Reply:
I just changed it. Now a garage door will open if closed and close if opened, even if a percentage is specified in the voice command.